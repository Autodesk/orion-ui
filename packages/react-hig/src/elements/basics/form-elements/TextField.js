/**
Copyright 2016 Autodesk,Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

*/
import React, { Component } from 'react';
import TextFieldAdapter from '../../../adapters/FormElements/TextFieldAdapter.js';
import PropTypes from 'prop-types';

class TextField extends Component {
  constructor(props) {
    super(props);

    const controlled = this.props.value === undefined ? false : true;

    this.state = {
      value: this.getDefaultValue(),
      controlled
    };
  }

  getDefaultValue() {
    const { defaultValue, value } = this.props;

    if (value !== undefined) {
      return value;
    } else if (defaultValue !== undefined) {
      return defaultValue;
    } else {
      return '';
    }
  }

  getRenderedValue() {
    const { value } = this.props;

    if (value !== undefined) {
      return value;
    } else {
      return this.state.value;
    }
  }

  handleInput = event => {
    if (this.props.onInput) {
      this.props.onInput(event);
    }

    if (this.state.controlled) {
      if (this.textFieldEl) {
        this.textFieldEl.forceNextReset();
        this.setState({ value: this.state.value });
      }
    } else {
      this.setState({ value: event.target.value });
    }
  };

  setTextFieldEl = textFieldEl => {
    this.textFieldEl = textFieldEl;
  };

  render() {
    return (
      <TextFieldAdapter
        ref={this.setTextFieldEl}
        {...this.props}
        value={this.getRenderedValue()}
        onInput={this.handleInput}
      />
    );
  }
}

TextField.propTypes = {
  disabled: PropTypes.bool,
  icon: PropTypes.string,
  instructions: PropTypes.string,
  label: PropTypes.string,
  name: PropTypes.string,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onInput: PropTypes.func,
  placeholder: PropTypes.string,
  required: PropTypes.string,
  value: PropTypes.string
};

TextField.__docgenInfo = {
  props: {
    disabled: {
      description: 'prevents interaction with the text field'
    },
    icon: {
      description: 'icon for the text field'
    },
    instructions: {
      description: 'instructional text for the text field'
    },
    label: {
      description: 'label for the text field'
    },
    name: {
      description: 'name of the field as submitted with a form'
    },
    onBlur: {
      description: 'callback called when user moves focus away from the text field'
    },
    onChange: {
      description: 'callback called when user changes the value of the text field and moves focus away'
    },
    onFocus: {
      description: 'callback called when user moves focus onto the text field'
    },
    onInput: {
      description: 'callback called when user changes the value of the text field'
    },
    placeholder: {
      description: 'text prompting the user to enter text'
    },
    required: {
      description: 'text indicating that the user must enter a value for this text field'
    }
  }
};

export default TextField;
