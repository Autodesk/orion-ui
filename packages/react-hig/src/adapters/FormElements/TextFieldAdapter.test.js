/**
 Copyright 2016 Autodesk,Inc.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.

 */
import { mount } from 'enzyme';
import * as HIG from 'hig.web';
import React from 'react';

import TextField from './TextFieldAdapter';

describe('<TextField>', () => {
  function createHigTextField(defaults = {}) {
    const higContainer = document.createElement('div');

    const higTextField = new HIG.TextField({ ...defaults });

    higTextField.mount(higContainer);

    return { higTextField, higContainer };
  }

  function createOrionTextField(props) {
    return <TextField {...props} />;
  }

  it('renders a text field', () => {
    const defaults = { name: 'mySpecialField' };

    const { higTextField, higContainer } = createHigTextField(defaults);
    const container = document.createElement('div');
    const wrapper = mount(createOrionTextField(defaults), {
      attachTo: container
    });

    expect(container.firstElementChild.outerHTML).toMatchSnapshot();

    expect(container.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('renders a text field with initial props', () => {
    const defaults = {
      icon: 'assets',
      instructions: "Don't just do something, sit there.",
      label: 'Name of your first pet',
      name: 'mySpecialField',
      placeholder: 'Was it Fluffy?',
      required: 'You really must fill this in.',
      value: 'Rex'
    };

    const { higTextField, higContainer } = createHigTextField(defaults);
    higTextField.required(defaults.required);
    const container = document.createElement('div');
    const wrapper = mount(createOrionTextField(defaults), {
      attachTo: container
    });

    expect(container.firstElementChild.outerHTML).toMatchSnapshot();

    expect(container.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('renders a text field with udpated props', () => {
    const defaults = {
      name: 'mySpecialField',
      disabled: true
    };

    const { higTextField, higContainer } = createHigTextField(defaults);
    const orionContainer = document.createElement('div');
    const wrapper = mount(createOrionTextField(defaults), {
      attachTo: orionContainer
    });

    const nextProps = {
      icon: 'assets',
      instructions: "Don't just do something, sit there.",
      label: 'Name of your first pet',
      placeholder: 'Was it Fluffy?',
      required: 'You really must fill this in.',
      value: 'Rex',
      disabled: false
    };

    higTextField.setIcon(nextProps.icon);
    higTextField.setInstructions(nextProps.instructions);
    higTextField.setLabel(nextProps.label);
    higTextField.setPlaceholder(nextProps.placeholder);
    higTextField.required(nextProps.required);
    higTextField.setValue(nextProps.value);
    higTextField.enable();
    wrapper.setProps(nextProps);

    expect(orionContainer.firstElementChild.outerHTML).toMatchSnapshot();

    expect(orionContainer.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  ['onBlur', 'onChange', 'onFocus', 'onInput'].forEach(eventName => {
    it(`sets event listeners for ${eventName} initially`, () => {
      const spy = jest.fn();
      const container = document.createElement('div');
      const wrapper = mount(createOrionTextField({ [eventName]: spy }), {
        attachTo: container
      });
      const instance = wrapper.instance().instance;

      const disposeFunction = instance._disposeFunctions.get(
        eventName + 'Dispose'
      );
      expect(disposeFunction).toBeDefined();
    });

    it(`sets event listeners for ${eventName} when updated`, () => {
      const spy = jest.fn();
      const container = document.createElement('div');
      const wrapper = mount(createOrionTextField({}), {
        attachTo: container
      });
      wrapper.setProps({ [eventName]: spy });

      const instance = wrapper.instance().instance;

      const disposeFunction = instance._disposeFunctions.get(
        `${eventName}Dispose`
      );
      expect(disposeFunction).toBeDefined();
    });
  });
});
