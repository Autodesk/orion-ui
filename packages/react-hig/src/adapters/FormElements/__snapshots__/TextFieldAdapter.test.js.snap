// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField> renders a text field 1`] = `
"<div class=\\"hig__text-field\\">
    <!--ICON-SPACER-->
    
    <div class=\\"hig__text-field__content\\">
        <div class=\\"hig__text-field__input-wrapper\\">
            <label class=\\"hig__text-field__icon\\" for=\\"text-field-mySpecialField\\"></label>
            <button class=\\"hig__text-field__extra hig__text-field__password-reveal-button\\"></button>
            <button class=\\"hig__text-field__extra hig__text-field__password-hide-button\\"></button>
            <!--EXTRA-->
            <!--SLOT-->
            <input name=\\"mySpecialField\\" id=\\"text-field-mySpecialField\\" class=\\"hig__text-field__input hig__text-field__input--no-value\\" placeholder=\\"\\" value=\\"\\">
            <label for=\\"text-field-mySpecialField\\" class=\\"hig__text-field__label \\"></label>
        </div>
        <!--INSTRUCTIONS-->
        
        <!--REQUIRED-NOTICE-->
    </div>
</div>"
`;

exports[`<TextField> renders a text field with initial props 1`] = `
"<div class=\\"hig__text-field hig__text-field--required\\">
    <!--ICON-SPACER-->
    <div class=\\"hig__text-field__icon-spacer\\"></div>
    <div class=\\"hig__text-field__content\\">
        <div class=\\"hig__text-field__input-wrapper\\">
            <label class=\\"hig__text-field__icon hig__text-field__icon--visible\\" for=\\"text-field-mySpecialField\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"assets\\"><title>hig-light/globalnav/sidenav/icon/assets</title><g fill=\\"#221F20\\" fill-rule=\\"evenodd\\"><path d=\\"M10.1898101,0.916614844 L9.67284723,8.01491201 L9.15588441,0.916614844 C9.15588441,0.916614844 7.37499587,-0.81644658 2.12744709,0.48294645 L2.12744709,12.0829042 C2.12744709,12.0829042 6.89296297,11.6014087 8.33744951,12.3725539 C8.33744951,12.3725539 8.29284669,12.9427177 9.66048741,12.950241 L9.67284723,12.950241 L9.68574443,12.950241 C11.0528478,12.9427177 11.0082449,12.3725539 11.0082449,12.3725539 C12.4527315,11.6014087 17.2182474,12.0829042 17.2182474,12.0829042 L17.2182474,0.48294645 C11.9706986,-0.81644658 10.1898101,0.916614844 10.1898101,0.916614844\\" transform=\\"translate(2.327 4.245)\\"></path><path d=\\"M0.91156344,2.55923464 L0.885231654,13.9028606 L0.455862327,13.4729539 L7.29245378,13.4192155 C7.34887904,13.421365 7.39186971,13.4697296 7.38972017,13.52508 C7.38649587,13.5777436 7.3435052,13.6196595 7.29245378,13.6218091 L0.455862327,14.3333047 C0.229623921,14.3440523 0.0372406687,14.1694027 0.0270303844,13.9442391 L0.0259556176,13.925968 L0.0259556176,13.9028606 L0.000161215016,2.55923464 C-0.000376168371,2.3082766 0.203292135,2.10353353 0.453712794,2.10245876 C0.705745602,2.101384 0.911026056,2.30612707 0.91156344,2.55708511 L0.91156344,2.55923464 Z\\" transform=\\"translate(2.327 4.245)\\"></path><path d=\\"M19.3458019,2.55923464 L19.3200075,13.9028606 L19.3200075,13.9248933 C19.3183954,14.1516691 19.1351477,14.3343794 18.9089092,14.3343794 L18.8895634,14.3333047 L12.0535094,13.6218091 C11.9970841,13.6196595 11.9540934,13.571295 11.956243,13.5159445 C11.9594673,13.463281 12.002458,13.421365 12.0535094,13.4192155 L18.8895634,13.4729539 L18.4607315,13.9028606 L18.4343997,2.55923464 C18.4333249,2.3082766 18.6375306,2.10353353 18.8879513,2.10245876 C19.1405215,2.101384 19.3452646,2.30612707 19.3458019,2.55708511 L19.3458019,2.55923464 Z\\" transform=\\"translate(2.327 4.245)\\"></path></g></svg></label>
            <button class=\\"hig__text-field__extra hig__text-field__password-reveal-button\\"></button>
            <button class=\\"hig__text-field__extra hig__text-field__password-hide-button\\"></button>
            <!--EXTRA-->
            <!--SLOT-->
            <input name=\\"mySpecialField\\" id=\\"text-field-mySpecialField\\" class=\\"hig__text-field__input\\" placeholder=\\"Was it Fluffy?\\" value=\\"Rex\\">
            <label for=\\"text-field-mySpecialField\\" class=\\"hig__text-field__label hig__text-field__label--visible\\">Name of your first pet</label>
        </div>
        <!--INSTRUCTIONS-->
        <p class=\\"hig__text-field__instructions\\">Don't just do something, sit there.</p>
        <p class=\\"hig__text-field__required-notice\\">You really must fill this in.</p><!--REQUIRED-NOTICE-->
    </div>
</div>"
`;

exports[`<TextField> renders a text field with udpated props 1`] = `
"<div class=\\"hig__text-field hig__text-field--required\\">
    <div class=\\"hig__text-field__icon-spacer\\"></div><!--ICON-SPACER-->
    
    <div class=\\"hig__text-field__content\\">
        <div class=\\"hig__text-field__input-wrapper\\">
            <label class=\\"hig__text-field__icon hig__text-field__icon--visible\\" for=\\"text-field-mySpecialField\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"assets\\"><title>hig-light/globalnav/sidenav/icon/assets</title><g fill=\\"#221F20\\" fill-rule=\\"evenodd\\"><path d=\\"M10.1898101,0.916614844 L9.67284723,8.01491201 L9.15588441,0.916614844 C9.15588441,0.916614844 7.37499587,-0.81644658 2.12744709,0.48294645 L2.12744709,12.0829042 C2.12744709,12.0829042 6.89296297,11.6014087 8.33744951,12.3725539 C8.33744951,12.3725539 8.29284669,12.9427177 9.66048741,12.950241 L9.67284723,12.950241 L9.68574443,12.950241 C11.0528478,12.9427177 11.0082449,12.3725539 11.0082449,12.3725539 C12.4527315,11.6014087 17.2182474,12.0829042 17.2182474,12.0829042 L17.2182474,0.48294645 C11.9706986,-0.81644658 10.1898101,0.916614844 10.1898101,0.916614844\\" transform=\\"translate(2.327 4.245)\\"></path><path d=\\"M0.91156344,2.55923464 L0.885231654,13.9028606 L0.455862327,13.4729539 L7.29245378,13.4192155 C7.34887904,13.421365 7.39186971,13.4697296 7.38972017,13.52508 C7.38649587,13.5777436 7.3435052,13.6196595 7.29245378,13.6218091 L0.455862327,14.3333047 C0.229623921,14.3440523 0.0372406687,14.1694027 0.0270303844,13.9442391 L0.0259556176,13.925968 L0.0259556176,13.9028606 L0.000161215016,2.55923464 C-0.000376168371,2.3082766 0.203292135,2.10353353 0.453712794,2.10245876 C0.705745602,2.101384 0.911026056,2.30612707 0.91156344,2.55708511 L0.91156344,2.55923464 Z\\" transform=\\"translate(2.327 4.245)\\"></path><path d=\\"M19.3458019,2.55923464 L19.3200075,13.9028606 L19.3200075,13.9248933 C19.3183954,14.1516691 19.1351477,14.3343794 18.9089092,14.3343794 L18.8895634,14.3333047 L12.0535094,13.6218091 C11.9970841,13.6196595 11.9540934,13.571295 11.956243,13.5159445 C11.9594673,13.463281 12.002458,13.421365 12.0535094,13.4192155 L18.8895634,13.4729539 L18.4607315,13.9028606 L18.4343997,2.55923464 C18.4333249,2.3082766 18.6375306,2.10353353 18.8879513,2.10245876 C19.1405215,2.101384 19.3452646,2.30612707 19.3458019,2.55708511 L19.3458019,2.55923464 Z\\" transform=\\"translate(2.327 4.245)\\"></path></g></svg></label>
            <button class=\\"hig__text-field__extra hig__text-field__password-reveal-button\\"></button>
            <button class=\\"hig__text-field__extra hig__text-field__password-hide-button\\"></button>
            <!--EXTRA-->
            <!--SLOT-->
            <input name=\\"mySpecialField\\" id=\\"text-field-mySpecialField\\" class=\\"hig__text-field__input\\" placeholder=\\"Was it Fluffy?\\" value=\\"Rex\\">
            <label for=\\"text-field-mySpecialField\\" class=\\"hig__text-field__label hig__text-field__label--visible\\">Name of your first pet</label>
        </div>
        <p class=\\"hig__text-field__instructions\\">Don't just do something, sit there.</p><!--INSTRUCTIONS-->
        
        <p class=\\"hig__text-field__required-notice\\">You really must fill this in.</p><!--REQUIRED-NOTICE-->
    </div>
</div>"
`;
